emma-flink/src/main/scala/org/emmalanguage/api/FlinkDataSet.scala

```diff
@@ -91,6 +97,47 @@ class FlinkDataSet[A: Meta] private[api]
     rep.distinct
 
   // -----------------------------------------------------
+  // Partition-based Ops
+  // -----------------------------------------------------
+
+  def sample(k: Int, seed: Long = 5394826801L): Vector[A] = {
+    // sample per partition and sorted by partition ID
+    val Seq(hd, tl@_*) = new DataSetUtils(rep).zipWithIndex
+      .mapPartition(new RichMapPartitionFunction[(Long, A), (Int, Array[Option[A]])] {
+        @transient private var pid = 0
+
+        override def open(parameters: Configuration) = {
+          super.open(parameters)
+          pid = getRuntimeContext.getIndexOfThisSubtask
+        }
+
+        import scala.collection.JavaConversions._
+
+        def mapPartition(it: lang.Iterable[(Long, A)], out: Collector[(Int, Array[Option[A]])]) = {
+          val sample = Array.fill(k)(Option.empty[A])
+          for ((i, e) <- it) {
+            if (i >= k) {
+              val j = util.RanHash(seed).at(i).nextLong(i + 1)
+              if (j < k) sample(j.toInt) = Some(e)
+            } else sample(i.toInt) = Some(e)
```

Comment:
I don't quite understand how this works. Won't this pick all values for the sample from the first partition if it has `>= k` items?